### ISSUE TITLE
#[repr(simd)] on structs vs arrays, motivation

### ISSUE THREAD

**Issue by gnzlbg** [rust-lang-nursery/stdsimd#37]
I have been trying to find the motivation for using `repr(simd)` on tuple structs instead of arrays for a while without any luck. Does anybody have some pointers? 

I must be missing something because to make using it on arrays makes much more sense, mainly because:

- we can use the recently merged type-level const RFC to create arbitrary sized simd vectors on stable (maybe preventing this was the aim of using tuple structs? if so, tuple structs will also fail here if we ever get variadics)
- we can use `[]` to index into the elements of a simd vector, which is much nicer than using .0 (maybe it is because `[]` always does bounds checking which would be bad for simd where a compile-time solution would be preferable?)

**Comment by BurntSushi** [comment]
I don't really have a dog in this fight. I defined them the way they are *a year ago* because that's how they were defined in the `simd` crate *many years ago*. :-) Any and all developments in Rust the language after that point had zero role in the design decision.

I haven't kept up with the type-level const RFC. I don't know it's implementation status. I think it would help if we could have some concrete code samples.

**Comment by gnzlbg** [comment]
Before we waste any time on this I am just going to ping @alexchrichton and
ask: was this discussed somewhere? And if so, do you have some links? I am
mostly curious, and since we are not going to pursue stabilizing repr[simd]
any time soon maybe this discussion is not worth having. We can manually
implement the Index traits on simd types if we just so desire.

**Comment by alexcrichton** [comment]
My guess is that it's just always been this way, I can't personally remember a time when we debated this and/or decided on structs.

Doing some digging through revealed what may be the [first implementation of SIMD in rust](https://github.com/rust-lang/rust/pull/5841) using dedicated syntax (not merged) followed by what I think is the [first merged PR for SIMD in Rust](https://github.com/rust-lang/rust/pull/6214). [This comment in particular](https://github.com/rust-lang/rust/pull/5841#issuecomment-16668732) may be of interest.

tl;dr; I don't think there's much to learn from the past here?

**Comment by gnzlbg** [comment]
I'm closing this. The comment you cited answered the question.
